srml/bridge/src/justification.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use client::{CallExecutor, Client};
+use client::backend::Backend;
+use client::error::Error as ClientError;
+use codec::{Encode, Decode};
+use core::cmp::{Ord, Ordering};
+use fg::{Commit, Error, Message}; // Q: Should I make any of these a part of fg_primitives?
+use fg_primitives::{AuthorityId, RoundNumber, SetId as SetIdNumber, AuthoritySignature};
+use grandpa::voter_set::VoterSet;
+use grandpa::{Error as GrandpaError};
+
+// Might be able to get this from primitives re-export
+use rstd::collections::{
+	btree_map::BTreeMap,
+	btree_set::BTreeSet,
+};
+
+use sr_primitives::app_crypto::RuntimeAppPublic;
+use sr_primitives::generic::BlockId;
+use sr_primitives::traits::{NumberFor, Block as BlockT, Header as HeaderT};
+use primitives::{H256, Blake2Hasher};
+
+/// A GRANDPA justification for block finality, it includes a commit message and
+/// an ancestry proof including all headers routing all precommit target blocks
+/// to the commit target block. Due to the current voting strategy the precommit
+/// targets should be the same as the commit target, since honest voters don't
+/// vote past authority set change blocks.
+///
+/// This is meant to be stored in the db and passed around the network to other
+/// nodes, and are used by syncing nodes to prove authority set handoffs.
+#[derive(Encode, Decode)]
+pub struct GrandpaJustification<Block: BlockT> {
```

Comment:
I feel like the contents of this module should be in the `pallette/grandpa` crate, since it's generally useful.

One thing - we have to deal with validator-set handoffs as well.

Reply:
IMO it should go into `primitives/finality-grandpa` and then it gets used by both client and pallet code (and also by the bridge pallet).

Reply:
I'm going to wait until I rebase past `the-big-reorg` before moving the code into `primitives/finality-grandpa`.